pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,103 @@
+
+	"github.com/gardener/gardener/pkg/apis/core"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+var (
+	specPath        = field.NewPath("spec")
+	nwPath          = specPath.Child("networking")
+	providerPath    = specPath.Child("provider")
+	infraConfigPath = providerPath.Child("infrastructureConfig")
+	cpConfigPath    = providerPath.Child("controlPlaneConfig")
+	workersPath     = providerPath.Child("workers")
+)
+
+func (v *Shoot) validateShoot(shoot *core.Shoot, infraConfig *azure.InfrastructureConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// ControlPlaneConfig
+	if shoot.Spec.Provider.ControlPlaneConfig != nil {
+		if _, err := decodeControlPlaneConfig(v.decoder, shoot.Spec.Provider.ControlPlaneConfig); err != nil {
+			allErrs = append(allErrs, field.Forbidden(cpConfigPath, "not allowed to configure an unsupported controlPlaneConfig"))
+		}
+	}
+
+	// Network validation
+	allErrs = append(allErrs, azurevalidation.ValidateNetworking(shoot.Spec.Networking, nwPath)...)
+
+	// Provider validation
+	allErrs = append(allErrs, azurevalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, infraConfigPath)...)
+
+	// Shoot workers
+	allErrs = append(allErrs, azurevalidation.ValidateWorkers(shoot.Spec.Provider.Workers, infraConfig.Zoned, workersPath)...)
+
+	return allErrs
+}
+
+func (v *Shoot) validateShootUpdate(oldShoot, shoot *core.Shoot) error {
+	// InfrastructureConfig update
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(infraConfigPath, "InfrastructureConfig must be set for Azure shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig)
+	if err != nil {
+		return field.Forbidden(infraConfigPath, "not allowed to configure an unsupported infrastructureConfig")
+	}
+
+	if oldShoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.InternalError(infraConfigPath, errors.New("InfrastructureConfig is not available on old shoot"))
+	}
+
+	oldInfraConfig, err := decodeInfrastructureConfig(v.decoder, oldShoot.Spec.Provider.InfrastructureConfig)
+	if err != nil {
+		return field.Forbidden(infraConfigPath, "not allowed to configure an unsupported infrastructureConfig")
+	}
+
+	allErrs := field.ErrorList{}
+
+	if !reflect.DeepEqual(oldInfraConfig, infraConfig) {
+		allErrs = append(allErrs, azurevalidation.ValidateInfrastructureConfigUpdate(oldInfraConfig, infraConfig, infraConfigPath)...)
+	}
+
+	allErrs = append(allErrs, azurevalidation.ValidateWorkersUpdate(oldShoot.Spec.Provider.Workers, shoot.Spec.Provider.Workers, workersPath)...)
+
+	allErrs = append(allErrs, v.validateShoot(shoot, infraConfig)...)
+
+	return allErrs.ToAggregate()
+}
+
+func (v *Shoot) validateShootCreation(shoot *core.Shoot) error {
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
```

Comment:
Why are lines 93-100 not moved into the `validateShoot` function as well? The `ControlPlaneConfig` related code is in this function as well.

Reply:
You need to do it before because in `validateShootUpdate` also the "old" object needs to be decoded + checked. If it were in `validateShoot ` we'd do it twice.